### PR TITLE
Add save-in capabilities for interface kernels

### LINKED ISSUE #9854: Add save_in for interface kernels
### Description of the enhancement or error report

Add `save_in` for interface kernels.

### Rationale for the enhancement or information for reproducing the error

Should be consistent with kernels and boundary conditions. Can be useful for debugging residuals.

### Identified impact

Consistency across primary objects that contribute to residuals and jacobians.

### PR DESCRIPTION
**Pull request by lindsayad**
A little more complicated than for other residual/jacobian objects because the interface kernel can contribute to two distinct variables residuals and diagonal Jacobians. Consequently we have:

- `_master_save_in_residual_variables`
- `_slave_save_in_residual_variables`
- `_master_save_in_jacobian_variables`
- `_slave_save_in_jacobian_variables`

Tests demonstrate zero combined residuals and correct diagonal Jacobians at the interface.

Closes #9854

### CODE REVIEW

File: framework/include/interfacekernels/InterfaceKernel.h [hunk 1]
```diff
@@ -53,6 +53,46 @@ class InterfaceKernel : public DGKernel
...
+  std::vector<MooseVariable *> _master_save_in_residual_variables;
+
+  /// Whether there are slave residual aux variables
+  bool _has_slave_residuals_saved_in;
+
+  /// The aux variables to save the slave contributions to
+  std::vector<MooseVariable *> _slave_save_in_residual_variables;
+
+  /// Vector of MooseEnums specifying whether jacobian save-in aux variables correspond to master or slave side
+  std::vector<MooseEnum> _diag_save_in_var_side;
```
Comment (permcody) [inline framework/include/interfacekernels/InterfaceKernel.h]: So I recommend that you don't create a vector of MooseEnums. The reason is that MooseEnums don't behave well if you try to push_back or extend the vector. We created `MultiMooseEnum` for this purpose instead.

File: framework/src/interfacekernels/InterfaceKernel.C [hunk 2]
```diff
@@ -27,20 +28,135 @@ validParams<InterfaceKernel>()
...
+      "The name of auxiliary variables to save this Kernel's residual contributions to. "
+      " Everything about that variable must match everything about this variable (the "
+      "type, what blocks it's on, etc.)");
+  params.addParam<std::vector<AuxVariableName>>(
+      "diag_save_in",
+      "The name of auxiliary variables to save this Kernel's diagonal Jacobian "
+      "contributions to. Everything about that variable must match everything "
+      "about this variable (the type, what blocks it's on, etc.)");
+
+  std::vector<MooseEnum> save_in_var_side(1, MooseEnum("m s"));
```
Comment (permcody) [inline framework/src/interfacekernels/InterfaceKernel.C]: Yeah, just kind of weird... Use `MultiMooseEnum`

File: framework/include/dgkernels/DGKernel.h [hunk 3]
```diff
@@ -222,7 +222,7 @@ class DGKernel : public MooseObject,
   // number)
   static const BoundaryID InternalBndId;
 
-private:
+protected:
```
Comment (permcody) [inline framework/include/dgkernels/DGKernel.h]: Just making sure. It's not wrong to use these same statics in two different classes but you'll want to verify that we don't evaluate DGKernels and Interface kernels at the same time (pretty sure we don't) or you might have a little bit of lock contention.

File: framework/src/interfacekernels/InterfaceKernel.C [hunk 4]
```diff
@@ -87,10 +223,38 @@ InterfaceKernel::computeElemNeighJacobian(Moose::DGJacobianType type)
...
     for (_i = 0; _i < test_space.size(); _i++)
       for (_j = 0; _j < loc_phi.size(); _j++)
-        Kxx(_i, _j) += _JxW[_qp] * _coord[_qp] * computeQpJacobian(type);
+        _local_kxx(_i, _j) += _JxW[_qp] * _coord[_qp] * computeQpJacobian(type);
+
+  Kxx += _local_kxx;
+
+  if (_has_master_jacobians_saved_in && type == Moose::ElementElement)
+  {
+    unsigned int rows = _local_kxx.m();
```
Comment (permcody) [inline framework/src/interfacekernels/InterfaceKernel.C]: Are you sure `m()` returns `unsigned int`? Good place for `auto` (catching the right return type every time).

**Review by permcody — CHANGES_REQUESTED** [review]
Looks mostly OK, There will be a few more changes that I need to review after you switch to `MultiMooseEnum`

File: framework/include/dgkernels/DGKernel.h (hunk 3 above)
Comment (lindsayad) [reply framework/include/dgkernels/DGKernel.h]: I want to make sure I'm understanding things. We have these locks to make sure that we don't have a race condition where multiple threads are trying to set the aux variable solution at the same time right? In the context of these save-in variables, you could be saving into the same variable from multiple kernel, dgkernel, bc, interface kernel objects. Wouldn't you want lock contention in that case? E.g. if a dgkernel was trying to add to the variable solution at the same time as an interface kernel was trying to add-in, wouldn't you want the lock in place?

Comment (permcody) [reply framework/include/dgkernels/DGKernel.h]: You've got it. If these are indeed the same vectors (as you just convinced me that they are), then we should be using the same locks. This is fine.

**Review by permcody — APPROVED** [review]

**Pull request merged**